slotsmanager.c

```diff
@@ -0,0 +1,427 @@
+
+    for(i = 0; i < NumPart; i++)
+        if(P[i].IsGarbage)
+        {
+            P[i] = P[NumPart - 1];
+
+            NumPart--;
+            i--;
+
+            count_elim++;
+        }
+
+    sumup_large_ints(1, &NumPart, &total);
+
+    if(total != total0) {
+        message(0, "GC : Reducing Particle slots from %ld to %ld\n", total0, total);
+        tree_invalid = 1;
+    }
+    return tree_invalid;
+}
+
+static int slot_cmp_reverse_link(const void * b1in, const void * b2in) {
+    const struct particle_data_ext * b1 = (struct particle_data_ext *) b1in;
+    const struct particle_data_ext * b2 = (struct particle_data_ext *) b2in;
+    if(b1->gc.ReverseLink == -1 && b2->gc.ReverseLink == -1) {
+        return 0;
+    }
+    if(b1->gc.ReverseLink == -1) return 1;
+    if(b2->gc.ReverseLink == -1) return -1;
+    return (b1->gc.ReverseLink > b2->gc.ReverseLink) - (b1->gc.ReverseLink < b2->gc.ReverseLink);
+
+}
+
+static int
+slots_gc_mark()
+{
+    int ptype, i;
+
+    for(ptype = 0; ptype < 6; ptype ++) {
+        if(!SlotsManager->info[ptype].enabled) continue;
+#pragma omp parallel for
+        for(i = 0; i < SlotsManager->info[ptype].size; i++) {
+            BASESLOT_PI(i, ptype)->gc.ReverseLink = -1;
+        }
+    }
+
+#pragma omp parallel for
+    for(i = 0; i < NumPart; i++) {
+        if(!SlotsManager->info[P[i].Type].enabled) continue;
+
+        BASESLOT(i)->gc.ReverseLink = i;
+    }
+}
+
+/* sweep removed unused elements. */
+static int
+slots_gc_sweep(int ptype)
+{
+    if(!SlotsManager->info[ptype].enabled) ;
+
+    int used = SlotsManager->info[ptype].size;
+    size_t elsize = SlotsManager->info[ptype].elsize;
+    int i = 0;
+    /* put unused guys to the end */
+    while(i < used)
+    {
+        while(i < used
+                &&
+        BASESLOT_PI(i, ptype)->gc.ReverseLink == -1) {
+
+            memcpy(BASESLOT_PI(i, ptype),
+                BASESLOT_PI(used - 1, ptype), elsize);
+            used -- ;
+        }
+        i ++;
+    }
+
+    SlotsManager->info[ptype].size = used;
+}
+
+/* defrags ensures locality. */
+static void
+slots_gc_defrag(int ptype)
+{
+
+    if(!SlotsManager->info[ptype].enabled) return;
+    /* measure the fragmentation */
+    int i;
+    int frag = 0;
+    for(i = 1;
+        i < SlotsManager->info[ptype].size;
+        i ++) {
+
+        if( BASESLOT_PI(i, ptype)->gc.ReverseLink <
+            BASESLOT_PI(i - 1, ptype)->gc.ReverseLink
+            ) {
+            frag++;
+        }
+    }
+    int defrag = MPIU_Any(
+            frag > SlotsManager->info[ptype].size * 0.1,
+            MPI_COMM_WORLD);
+
+    /* if any rank is too fragmented, add a sorting to defrag. */
+    if(defrag)  {
+        /* sort the used ones
+         * by their location in the P array */
+        qsort_openmp(SlotsManager->info[ptype].ptr,
+                     SlotsManager->info[ptype].size,
+                     SlotsManager->info[ptype].elsize, 
+                     slot_cmp_reverse_link);
```

Comment:
Looks good. I did some profiling (perhaps you did too) and qsorting the SPH slots was taking 90% of the GC time. So avoiding this if defrag is too high is a good idea. We should probably profile this a bit though: if the sort is run too rarely the tree will become slow (and if too often the gc takes too much time). But we can mess with that later.

Reply:
Yes. Later. I was thinking adding a separate defrag that is called only in full domain decomposition.